saleor/discount/migrations/tasks/saleor3_16.py

```diff
@@ -253,26 +281,57 @@ def migrate_checkout_line_discounts(sales_pks, rule_by_channel_and_sale):
 
 
 def migrate_order_line_discounts(sales_pks, rule_by_channel_and_sale):
-    # TODO: to update - should be created based on `OrderLine.sale_id`
-    if order_line_discounts := OrderLineDiscount.objects.filter(
-        sale_id__in=sales_pks
-    ).select_related("line__order"):
-        for order_line_discount in order_line_discounts:
-            if order_line := order_line_discount.line:
-                channel_id = order_line.order.channel_id
-                sale_id = order_line_discount.sale_id
-                lookup = f"{channel_id}_{sale_id}"
-                order_line_discount.type = "promotion"
-                if promotion_rule := rule_by_channel_and_sale.get(lookup):
-                    order_line_discount.promotion_rule = promotion_rule
+    global_pks = [graphene.Node.to_global_id("Sale", pk) for pk in sales_pks]
+    if order_lines := OrderLine.objects.filter(sale_id__in=global_pks).select_related(
+        "order"
+    ):
+        order_line_discounts = []
+        for order_line in order_lines:
+            channel_id = order_line.order.channel_id
+            sale_id = graphene.Node.from_global_id(order_line.sale_id)[1]
+            lookup = f"{channel_id}_{sale_id}"
+            if rule := rule_by_channel_and_sale.get(lookup):
+                order_line_discounts.append(
+                    OrderLineDiscount(
+                        type="promotion",
+                        value_type=rule.reward_value_type,
+                        value=rule.reward_value,
+                        amount_value=get_discount_amount_value(order_line),
+                        currency=order_line.currency,
+                        promotion_rule=rule,
+                        line=order_line,
+                    )
+                )
 
-        OrderLineDiscount.objects.bulk_update(
-            order_line_discounts, ["promotion_rule_id", "type"]
-        )
+        OrderLineDiscount.objects.bulk_create(order_line_discounts)
+
+
+def get_discount_amount_value(order_line):
+    precision = get_currency_precision(order_line.currency)
+    number_places = Decimal(10) ** -precision
+    price = order_line.quantity * order_line.unit_discount_amount
+    return price.quantize(number_places)
 
 
 def get_rule_by_channel_sale(rules_info):
     return {
         f"{rule_info.channel_id}_{rule_info.sale_id}": rule_info.rule
         for rule_info in rules_info
     }
+
+
+@app.task
+def update_discounted_prices_task(
+    start_pk: int = 0,
+):
+    products = list(
+        Product.objects.filter(pk__gt=start_pk)
+        .prefetch_related("channel_listings", "collections")
+        .order_by("pk")[:DISCOUNTED_PRICES_RECALCULATION_BATCH_SIZE]
+    )
+
+    if products:
+        update_products_discounted_price(products)
```

Comment:
Correct me if I'm wrong, but doesn't we fetch products again in this method. If so, there is no need to fetch all data about products here.